src/bot/exts/rdap.py

```diff
@@ -0,0 +1,219 @@
+        """Finds the first entity with the specified role."""
+        for entity in self.entities:
+            if role in entity.roles:
+                return entity
+        return None
+
+
+class RDAPDomain(RDAPResponse):
+    """Model for Domain RDAP responses."""
+
+    ldhName: str | None = None
+    events: list[dict[str, Any]] = []
+    nameservers: list[dict[str, Any]] = []
+
+    @property
+    def registration_date(self) -> str | None:
+        """Extracts the registration date from events."""
+        for event in self.events:
+            if event.get("eventAction") == "registration":
+                return event.get("eventDate")
+        return None
+
+
+class RDAPIP(RDAPResponse):
+    """Model for IP Network RDAP responses."""
+
+    startAddress: str | None = None
+    endAddress: str | None = None
+    name: str | None = None
+    parentHandle: str | None = None
+    type: str | None = None
+
+
+class RDAPASN(RDAPResponse):
+    """Model for Autonomous System Number RDAP responses."""
+
+    startAutnum: int | None = None
+    endAutnum: int | None = None
+    name: str | None = None
+    type: str | None = None
+
+
+class RDAP(commands.Cog):
+    """RDAP lookup commands."""
+
+    def __init__(self, bot: Bot):
+        self.bot = bot
+
+    def _format_table(self, data: dict[str, Any]) -> str:
+        """Format a dictionary as a markdown table."""
+        if not data:
+            return "No data available."
+
+        clean_data = {k: v for k, v in data.items() if v is not None}
+
+        if not clean_data:
+            return "No data available."
+
+        max_key_len = max(len(k) for k in clean_data)
+        lines: list[str] = []
+
+        lines.append(f"{'Property'.ljust(max_key_len)} | Value")
+        lines.append(f"{'-' * max_key_len}-|{'-' * 25}")
+
+        for key, value in clean_data.items():
+            lines.append(f"{key.ljust(max_key_len)} | {value}")
+
+        return "```\n" + "\n".join(lines) + "\n```"
+
+    @commands.command(name="rdap")
+    async def rdap_command(self, ctx: commands.Context[Bot], query: str) -> None:
+        """
+        Perform an RDAP lookup for a domain, IP, or ASN.
+
+        Usage:
+        !rdap example.com
+        !rdap 1.1.1.1
+        !rdap AS13335
+        """
+        query_type = classify_query(query)
+        url = f"{BaseURLs.rdap}/{query_type}/{query}"
+
+        try:
+            async with self.bot.http_session.get(url) as response:
+                if response.status == 404:
+                    await ctx.send(f"❌ No results found for `{query}`.")
+                    return
+                if response.status != 200:
+                    log.warning(f"RDAP lookup failed for {query}: HTTP {response.status}")
+                    await ctx.send(f"❌ Error fetching RDAP data: HTTP {response.status}")
+                    return
+
+                data = await response.json()
+        except Exception as e:
```

Comment:
this except block will never go off for RDAP errors because by default. aiohttp doesn't raise errors. you most likely want to handle this as the `else:` case for the above if-statements